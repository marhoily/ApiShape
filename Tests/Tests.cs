using System;
using System.Collections.Generic;
using System.Net.NetworkInformation;
using ApiShape;
using ApprovalTests;
using ApprovalTests.Reporters;
using Xunit;

namespace Tests
{
    public abstract class Abstract : IInvisible
    {
    }
    [Flags]
    public enum BigFlags
    {
        V0 = 0x1,
        V1 = 0x10,
        V2 = 0x100,
        V3 = 0x1000
    }
    public enum Byte : byte
    {
    }
    public enum Default
    {
    }
    public sealed class DeriveAndImpl : Impl
    {
    }
    public sealed class Derived : Abstract
    {
    }
    public sealed class DeriveLong : Generic<List<List<List<List<List<List<List<List<int>>>>>>>>>
    {
    }
    [Flags]
    public enum Flags
    {
    }
    public class Generic<T>
    {
    }
    public delegate T1 GenericArgsAndReturnValue<out T1, in T2>(T2 a1 = default(T2));
    public delegate void GenericConstraints<T>()
        where T : Abstract, new();
    public struct GenericImplGeneric<T> : IGeneric<T>
    {
    }
    public sealed class GenericWithConstraint<T> 
        where T : IUsual
    {
    }
    public interface IContravariant<in T>
    {
    }
    public interface ICovariant<out T>
    {
    }
    public interface IDerived : IUsual
    {
    }
    public interface IGeneric<T>
    {
    }
    public interface IGenericWithConstraint<T> 
        where T : IUsual
    {
    }
    public interface IIndirectDerive : IDerived
    {
        void F();
    }
    public interface IM
    {
        int P1 { get; }
        int P2 { get; set; }
        int P3 { set; }
        void M1();
        void M2(int a1);
        void M3(int a1, int a2);
        int M4();
        int M5<T>();
        int M6<T>() 
            where T : IUsual;
    }
    public class Impl : IUsual
    {
    }
    public sealed class ImplAndDeriveLong : Generic<List<List<List<List<List<List<List<List<int>>>>>>>>>, IGeneric<List<List<List<List<List<List<List<List<int>>>>>>>>>
    {
    }
    public struct ImplGeneric : IGeneric<int>
    {
    }
    public sealed class ImplLong : IGeneric<List<List<List<List<List<List<List<List<int>>>>>>>>>
    {
    }
    public interface IMultiDerived : IDerived, IGeneric<int>
    {
    }
    public sealed class IndirectImpl : Impl
    {
    }
    public struct InheritConstraint<T> : IGenericWithConstraint<T>
        where T : IUsual
    {
    }
    public interface IUsual
    {
    }
    public abstract class M
    {
        public int F1;
        public readonly int F2;
        protected int F3;
        public static int F4;
        public Uri F5;
        public const int X1 = 0;
        public const string X2 = "blah";
        public const string X3 = null;
        public const double X4 = 0.0;
        public const double X5 = 15e9;
        public int P1 { get; }
        public int P2 { get; set; }
        public int P3 { get; protected set; }
        public Uri P4 { get; private set; }
        public int P5 { private get; set; }
        public int P6 { protected get; set; }
        protected int P7 { get; }
        protected int P8 { get; set; }
        protected int P9 { private get; set; }
        protected int P10 { get; private set; }
        public abstract int P11 { get; }
        public virtual int P12 { get; }
        public static int P13 { get; }
        public string this[int index] { get { return ""; } set { } }
        protected string this[string index] { get { return ""; } set { } }
        public abstract string this[double index] { get; set; }
        public virtual string this[double index, int arg2] { get { return ""; } set { } }
        public abstract string this[byte i] { protected get; set; }
        public virtual string this[short i] { private get { return ""; } set { } }
        public virtual string this[ushort i] => "";
        protected M() { }
        public M(int a1) { }
        public abstract void M1();
        public abstract void M2(int a1);
        public abstract void M3(int a1, int a2);
        public abstract int M4();
        public abstract int M5<T>();
        public abstract int M6<T>() where T : IUsual;
        public virtual void M7() { }
        public void M8() { }
        protected abstract void M9();
        protected abstract void M10(int a1);
        protected abstract void M11(Uri a1, int a2);
        protected abstract int M12();
        protected abstract int M13<T>();
        protected abstract int M14<T>() where T : IUsual;
        protected virtual void M15() { }
        protected void M16() { }
        public abstract void M17(ref int a1, out int a2);
        public abstract void M28(ref int longLongLongLongLongLongLongLongArg1, out int longLongLongLongLongLongLongLongArg2);
        public static void M18() { }
        public abstract int M19<T>() where T : class;
        public abstract int M20<T>() where T : struct;
        public abstract int M21<T>() where T : new();
        public abstract int M28<T>() 
            where T : class, IGeneric<int>, new();
        public abstract int M22<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>();
        public abstract int M23<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>(TLongLongLongLongLongArg1 a1, TLongLongLongLongLongArg2 a2, TLongLongLongLongLongArg3 a3);
        public abstract int M24(string longLongLongName, Dictionary<string, List<Uri>> anotherLongName, string notLongEnough);
        public abstract int M25(string arg = null);
        public abstract int M29(int a1, int a2, string longLongLongLongLongLongLongLongArg1 = null);
        public abstract int M26(params string[] arg);
        public abstract int M27<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>()
            where TLongLongLongLongLongArg1 : ICloneable;
        public event EventHandler E1;
        public event EventHandler<NetworkAvailabilityEventArgs> E2 { add { } remove { } }
        public event Action E3;
        protected event Action E4;
        public sealed class NestedPublic
        {
        }
        protected class NestedProtected
        {
        }
        internal class NestedInternal
        {
        }
        private class NestedPrivate
        {
        }
    }
    public abstract class MM : M
    {
        private MM() { }
        private int F1;
        private const int X1 = 0;
        private string this[int index] => "";
        private void M27(params string[] arg) { }
        private event Action E5;

        protected MM(int f1) { F1 = f1; }
        protected MM(int a1, int f1) : base(a1) { F1 = f1; }
        public override int P11 { get; }
        public override int P12 { get; }
        public override string this[double index] { get { return ""; } set {  } }
        public override void M1() { }
    }
    public class NonSealed
    {
    }
    public delegate void OutAndRefAndParams(out int a1, ref string a2, params object[] a3);
    public delegate int ReturnValueAndArg(int a1, int opt = 0);
    public delegate void SampleDelegate();
    public sealed class Sealed
    {
    }
    public enum Seq
    {
        V0,
        V1,
        V2
    }
    public enum Short : short
    {
    }
    public static class Static
    {
    }
    public struct Struct
    {
    }
    public struct StructContainer : IIndirectDerive
    {
        public int F1;
        public readonly int F2;
        public static int F4;
        public const int X1 = 0;
        public int P1 { get; }
        public static int P13 { get; }
        public int P2 { get; set; }
        public int P4 { get; private set; }
        public int P5 { private get; set; }
        public StructContainer(int a1) { F1 = 0; F2 = 0; P1 = 0; P2 = 0; P4 = 0; P5 = 0; }
        public void F() { }
        public void M1() { }
        public static void M18() { }
        public void M2(int a1) { }
        public void M3(int a1, int a2) { }
        public int M4() => 3;
        public int M5<T>() => 3;
        public int M6<T>() 
			where T : IUsual => 3;
        public override string ToString() => "";
        public string this[int index] { get { return ""; } set { } }
        public string this[float index] { private get { return ""; } set { } }
        public string this[sbyte index] { get { return ""; } private set { } }
        public string this[double index] { set { } }
        public string this[byte index] => "";
        public static implicit operator StructContainer(int x) => new StructContainer();
        public static explicit operator int(StructContainer x) => 5;
        public static int operator +(StructContainer x, int i) => 5;
    }
    public struct StructImpl : IUsual
    {
    }
    public enum ULong : ulong
    {
    }
    [UseReporter(typeof(VisualStudioReporter))]
    public sealed class Tests
    {
        [Fact]
        public void All() => Approvals.Verify(typeof(Tests).Assembly.GetShape());
    }
    internal interface IInvisible { }
    internal struct Invisible { }
}