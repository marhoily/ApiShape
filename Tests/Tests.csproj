<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{C6FCC086-66CD-4FB8-8227-1B1DF142F14B}</ProjectGuid>
    <OutputType>Library</OutputType>
    <AppDesignerFolder>Properties</AppDesignerFolder>
    <RootNamespace>Tests</RootNamespace>
    <AssemblyName>Tests</AssemblyName>
    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <OutputPath>bin\Release\</OutputPath>
    <DefineConstants>TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <ItemGroup>
    <Reference Include="ApprovalTests, Version=3.0.0.0, Culture=neutral, PublicKeyToken=11bd7d124fc62e0f, processorArchitecture=MSIL">
      <HintPath>..\packages\ApprovalTests.3.0.11\lib\net40\ApprovalTests.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="ApprovalUtilities, Version=3.0.0.0, Culture=neutral, PublicKeyToken=11bd7d124fc62e0f, processorArchitecture=MSIL">
      <HintPath>..\packages\ApprovalUtilities.3.0.11\lib\net45\ApprovalUtilities.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="ApprovalUtilities.Net45, Version=3.0.0.0, Culture=neutral, processorArchitecture=MSIL">
      <HintPath>..\packages\ApprovalUtilities.3.0.11\lib\net45\ApprovalUtilities.Net45.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="FluentAssertions, Version=4.13.0.0, Culture=neutral, PublicKeyToken=33f2691a05b67b6a, processorArchitecture=MSIL">
      <HintPath>..\packages\FluentAssertions.4.13.0\lib\net45\FluentAssertions.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="FluentAssertions.Core, Version=4.13.0.0, Culture=neutral, PublicKeyToken=33f2691a05b67b6a, processorArchitecture=MSIL">
      <HintPath>..\packages\FluentAssertions.4.13.0\lib\net45\FluentAssertions.Core.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="System" />
    <Reference Include="System.Core" />
    <Reference Include="System.Xml.Linq" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.Data" />
    <Reference Include="System.Net.Http" />
    <Reference Include="System.Xml" />
    <Reference Include="xunit.abstractions, Version=2.0.0.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
      <HintPath>..\packages\xunit.abstractions.2.0.0\lib\net35\xunit.abstractions.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="xunit.assert, Version=2.1.0.3179, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
      <HintPath>..\packages\xunit.assert.2.1.0\lib\dotnet\xunit.assert.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="xunit.core, Version=2.1.0.3179, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
      <HintPath>..\packages\xunit.extensibility.core.2.1.0\lib\dotnet\xunit.core.dll</HintPath>
      <Private>True</Private>
    </Reference>
    <Reference Include="xunit.execution.desktop, Version=2.1.0.3179, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c, processorArchitecture=MSIL">
      <HintPath>..\packages\xunit.extensibility.execution.2.1.0\lib\net45\xunit.execution.desktop.dll</HintPath>
      <Private>True</Private>
    </Reference>
  </ItemGroup>
  <ItemGroup>
    <Compile Include="FormatClass.cs" />
    <Compile Include="Properties\AssemblyInfo.cs" />
  </ItemGroup>
  <ItemGroup>
    <None Include="packages.config" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\ApiShape\ApiShape.csproj">
      <Project>{5c017183-c38b-45d6-974a-5195c06c8204}</Project>
      <Name>ApiShape</Name>
    </ProjectReference>
  </ItemGroup>
  <ItemGroup>
    <Content Include="FormatClass.C1.approved.txt" />
    <Content Include="FormatClass.C2.approved.txt" />
    <Content Include="FormatClass.C3.approved.txt" />
    <Content Include="FormatClass.C4.approved.txt" />
    <Content Include="FormatClass.C5.approved.txt" />
    <Content Include="FormatClass.C6.approved.txt" />
    <Content Include="FormatClass.C7.approved.txt" />
    <Content Include="FormatClass.C8.approved.txt" />
    <Content Include="FormatClass.C9.approved.txt" />
    <Content Include="FormatClass.E1.approved.txt" />
    <Content Include="FormatClass.E2.approved.txt" />
    <Content Include="FormatClass.E3.approved.txt" />
    <Content Include="FormatClass.E4.approved.txt" />
    <Content Include="FormatClass.E5.approved.txt" />
    <Content Include="FormatClass.E6.approved.txt" />
    <Content Include="FormatClass.E7.approved.txt" />
    <Content Include="FormatClass.F1.approved.txt" />
    <Content Include="FormatClass.F2.approved.txt" />
    <Content Include="FormatClass.F3.approved.txt" />
    <Content Include="FormatClass.F4.approved.txt" />
    <Content Include="FormatClass.I1.approved.txt" />
    <Content Include="FormatClass.I2.approved.txt" />
    <Content Include="FormatClass.I3.approved.txt" />
    <Content Include="FormatClass.I4.approved.txt" />
    <Content Include="FormatClass.I5.approved.txt" />
    <Content Include="FormatClass.I6.approved.txt" />
    <Content Include="FormatClass.I7.approved.txt" />
    <Content Include="FormatClass.Im1.approved.txt" />
    <Content Include="FormatClass.Im2.approved.txt" />
    <Content Include="FormatClass.Im3.approved.txt" />
    <Content Include="FormatClass.Im4.approved.txt" />
    <Content Include="FormatClass.Im5.approved.txt" />
    <Content Include="FormatClass.Im6.approved.txt" />
    <Content Include="FormatClass.Ip1.approved.txt" />
    <Content Include="FormatClass.Ip2.approved.txt" />
    <Content Include="FormatClass.Ip3.approved.txt" />
    <Content Include="FormatClass.M1.approved.txt" />
    <Content Include="FormatClass.M10.approved.txt" />
    <Content Include="FormatClass.M11.approved.txt" />
    <Content Include="FormatClass.M12.approved.txt" />
    <Content Include="FormatClass.M13.approved.txt" />
    <Content Include="FormatClass.M14.approved.txt" />
    <Content Include="FormatClass.M15.approved.txt" />
    <Content Include="FormatClass.M16.approved.txt" />
    <Content Include="FormatClass.M17.approved.txt" />
    <Content Include="FormatClass.M2.approved.txt" />
    <Content Include="FormatClass.M3.approved.txt" />
    <Content Include="FormatClass.M4.approved.txt" />
    <Content Include="FormatClass.M5.approved.txt" />
    <Content Include="FormatClass.M6.approved.txt" />
    <Content Include="FormatClass.M7.approved.txt" />
    <Content Include="FormatClass.M8.approved.txt" />
    <Content Include="FormatClass.M9.approved.txt" />
    <Content Include="FormatClass.P1.approved.txt" />
    <Content Include="FormatClass.P10.approved.txt" />
    <Content Include="FormatClass.P2.approved.txt" />
    <Content Include="FormatClass.P3.approved.txt" />
    <Content Include="FormatClass.P4.approved.txt" />
    <Content Include="FormatClass.P5.approved.txt" />
    <Content Include="FormatClass.P6.approved.txt" />
    <Content Include="FormatClass.P7.approved.txt" />
    <Content Include="FormatClass.P8.approved.txt" />
    <Content Include="FormatClass.P9.approved.txt" />
    <Content Include="FormatClass.S1.approved.txt" />
    <Content Include="FormatClass.S2.approved.txt" />
    <Content Include="FormatClass.S3.approved.txt" />
    <Content Include="FormatClass.S4.approved.txt" />
    <Content Include="FormatClass.S5.approved.txt" />
    <Content Include="FormatClass.X1.approved.txt" />
    <Content Include="FormatClass.X2.approved.txt" />
    <Content Include="FormatClass.X3.approved.txt" />
    <Content Include="FormatClass.X4.approved.txt" />
    <Content Include="FormatClass.X5.approved.txt" />
  </ItemGroup>
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
       Other similar extension points exist, see Microsoft.Common.targets.
  <Target Name="BeforeBuild">
  </Target>
  <Target Name="AfterBuild">
  </Target>
  -->
</Project>