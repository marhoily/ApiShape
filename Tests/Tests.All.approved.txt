Full name: Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
Image runtime version: v4.0.30319
abstract class Tests.Abstract
{
}
[Flags]
enum Tests.BigFlags : int
{
    V0 = 1
    V1 = 16
    V2 = 256
    V3 = 4096
}
enum Tests.Byte : byte
{
}
enum Tests.Default : int
{
}
class Tests.DeriveAndImpl : Tests.Impl, Tests.IUsual
{
    constructor();
}
class Tests.Derived : Tests.Abstract
{
    constructor();
}
class Tests.DeriveLong : Tests.Generic<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<int>>>>>>>>>
{
    constructor();
}
[Flags]
enum Tests.Flags : int
{
}
virtual class Tests.Generic<T>
{
    constructor();
}
delegate T1 Tests.GenericArgsAndReturnValue<out T1, in T2>(T2 a1 = default(T2));
delegate void Tests.GenericConstraints<T>()
    where T : Abstract, new();
struct Tests.GenericImplGeneric<T> : IGeneric<T>
{
}
class Tests.GenericWithConstraint<T>
    where T : IUsual
{
    constructor();
}
interface Tests.IContravariant<in T>
{
}
interface Tests.ICovariant<out T>
{
}
interface Tests.IDerived : Tests.IUsual
{
}
interface Tests.IGeneric<T>
{
}
interface Tests.IGenericWithConstraint<T>
    where T : IUsual
{
}
interface Tests.IIndirectDerive : Tests.IDerived, Tests.IUsual
{
    void F();
}
interface Tests.IM
{
    int P1 { get; }
    int P2 { get; set; }
    int P3 { set; }
    void M1();
    void M2(int a1);
    void M3(int a1, int a2);
    int M4();
    int M5<T>();
    int M6<T>()
        where T : IUsual;
}
virtual class Tests.Impl : Tests.IUsual
{
    constructor();
}
class Tests.ImplAndDeriveLong : Tests.Generic<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<int>>>>>>>>>, Tests.IGeneric<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<int>>>>>>>>>
{
    constructor();
}
struct Tests.ImplGeneric : Tests.IGeneric<int>
{
}
class Tests.ImplLong : Tests.IGeneric<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<int>>>>>>>>>
{
    constructor();
}
interface Tests.IMultiDerived : Tests.IDerived, Tests.IGeneric<int>, Tests.IUsual
{
}
class Tests.IndirectImpl : Tests.Impl, Tests.IUsual
{
    constructor();
}
struct Tests.InheritConstraint<T> : IGenericWithConstraint<T>
    where T : IUsual
{
}
interface Tests.IUsual
{
}
abstract class Tests.M
{
    int F1;
    readonly int F2;
    int F3;
    static int F4;
    const int X1;
    const string X2;
    const string X3;
    const double X4;
    const double X5;
    string this[int index] { get; set; }
    protected string this[string index] { get; set; }
    string this[double index] { get; set; }
    string this[double index, int arg2] { get; set; }
    string this[byte i] { protected get; set; }
    string this[short i] { set; }
    string this[ushort i] { get; }
    int P1 { get; }
    int P10 { protected get; }
    int P11 { get; }
    int P12 { get; }
    int P13 { get; }
    int P2 { get; set; }
    int P3 { get; protected set; }
    int P4 { get; }
    int P5 { set; }
    int P6 { protected get; set; }
    protected int P7 { get; }
    protected int P8 { get; set; }
    int P9 { protected set; }
    constructor(int a1);
    abstract void M1();
    protected abstract void M10(int a1);
    protected abstract void M11(int a1, int a2);
    protected abstract int M12();
    protected abstract int M13<T>();
    protected abstract int M14<T>()
        where T : IUsual;
    protected virtual void M15();
    protected void M16();
    abstract void M17(ref int a1, out int a2);
    void M18();
    abstract int M19<T>();
    abstract void M2(int a1);
    abstract int M20<T>()
        where T : struct;
    abstract int M21<T>();
    abstract int M22<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>();
    abstract int M23<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>(TLongLongLongLongLongArg1 a1, TLongLongLongLongLongArg2 a2, TLongLongLongLongLongArg3 a3);
    abstract int M24(Dictionary<string, List<Uri>> anotherLongName, string longLongLongName, string notLongEnough);
    abstract int M25(string arg = default(string));
    abstract int M26(params string[] arg);
    abstract int M27<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>()
        where TLongLongLongLongLongArg1 : ICloneable;
    abstract void M28(ref int longLongLongLongLongLongLongLongArg1, out int longLongLongLongLongLongLongLongArg2);
    abstract int M28<T>()
        where T : class, IGeneric<int>, new();
    abstract int M29(int a1, int a2, string longLongLongLongLongLongLongLongArg1 = default(string));
    abstract void M3(int a1, int a2);
    abstract int M4();
    abstract int M5<T>();
    abstract int M6<T>()
        where T : IUsual;
    virtual void M7();
    void M8();
    protected abstract void M9();
    event EventHandler E1;
    event EventHandler<NetworkAvailabilityEventArgs> E2;
    event Action E3;
}
class Tests.M+NestedPublic
{
    constructor();
}
abstract class Tests.MM : Tests.M
{
    int F1;
    readonly int F2;
    int F3;
    string this[double index] { get; set; }
    int P11 { get; }
    int P12 { get; }
    virtual void M1();
    event EventHandler E1;
    event EventHandler<NetworkAvailabilityEventArgs> E2;
    event Action E3;
}
virtual class Tests.NonSealed
{
    constructor();
}
delegate void Tests.OutAndRefAndParams(out int a1, ref string a2, params object[] a3);
delegate int Tests.ReturnValueAndArg(int a1, int opt = 0);
delegate void Tests.SampleDelegate();
class Tests.Sealed
{
    constructor();
}
enum Tests.Seq : int
{
    V0 = 0
    V1 = 1
    V2 = 2
}
enum Tests.Short : short
{
}
abstract class Tests.Static
{
}
struct Tests.Struct
{
}
struct Tests.StructContainer : Tests.IDerived, Tests.IIndirectDerive, Tests.IUsual
{
    int F1;
    readonly int F2;
    static int F4;
    const int X1;
    string this[int index] { get; set; }
    string this[float index] { set; }
    string this[sbyte index] { get; }
    string this[double index] { set; }
    string this[byte index] { get; }
    int P1 { get; }
    int P13 { get; }
    int P2 { get; set; }
    int P4 { get; }
    int P5 { set; }
    constructor(int a1);
    virtual void F();
    void M1();
    void M18();
    void M2(int a1);
    void M3(int a1, int a2);
    int M4();
    int M5<T>();
    int M6<T>()
        where T : IUsual;
    virtual string ToString();
}
struct Tests.StructImpl : Tests.IUsual
{
}
class Tests.Tests
{
    constructor();
    void All();
}
enum Tests.ULong : ulong
{
}
